### PR TITLE
`pipeline` support for `device="mps"` (or any other string)

### PR DESCRIPTION
**Pull request by julien-c**
No tests yet given we don't officially support `torch==1.12` yet (and M1-based CI is still a work in progress)

### CODE REVIEW

File: src/transformers/pipelines/base.py [hunk 1]
```diff
@@ -763,11 +763,15 @@ def __init__(
         if is_torch_available() and isinstance(device, torch.device):
             self.device = device
         else:
-            self.device = device if framework == "tf" else torch.device("cpu" if device < 0 else f"cuda:{device}")
+            self.device = (
+                device
+                if framework == "tf"
+                else torch.device(device if type(device) == str else "cpu" if device < 0 else f"cuda:{device}")
+            )
```
Comment (sgugger) [inline src/transformers/pipelines/base.py:770]: Those are too many if and else for my little brain. Can we maybe do more if/elif/else ?

**Review by sgugger — APPROVED** [review]
Looking nice, thanks a lot!

**Comment by AsaKal** [comment]
Thanks @julien-c .  am getting a new error
My guess this is on the pytorch op.
> NotImplementedError: The operator 'aten::unique_consecutive' is not current implemented for the MPS device. If you want this op to be added in priority during the prototype phase of this feature, please comment on https://github.com/pytorch/pytorch/issues/77764. As a temporary fix, you can set the environment variable `PYTORCH_ENABLE_MPS_FALLBACK=1` to use the CPU as a fallback for this op. WARNING: this will be slower than running natively on MPS.



Here is the traceback

> preds = pipe('i am feeling awesome', ['positive', 'negative'])
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/pipelines/zero_shot_classification.py", line 182, in __call__
>     return super().__call__(sequences, **kwargs)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1071, in __call__
>     return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1093, in run_single
>     model_outputs = self.forward(model_inputs, **forward_params)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/pipelines/base.py", line 987, in forward
>     model_outputs = self._forward(model_inputs, **forward_params)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/pipelines/zero_shot_classification.py", line 201, in _forward
>     outputs = self.model(**model_inputs)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1186, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/transformers/models/bart/modeling_bart.py", line 1516, in forward
>     if len(torch.unique_consecutive(eos_mask.sum(1))) > 1:
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/torch/_jit_internal.py", line 447, in fn
>     return if_false(*args, **kwargs)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/torch/_jit_internal.py", line 447, in fn
>     return if_false(*args, **kwargs)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/torch/functional.py", line 913, in _consecutive_return_output
>     output, _, _ = _unique_consecutive_impl(input, return_inverse, return_counts, dim)
>   File "/Volumes/training/torch-gpu/env/lib/python3.8/site-packages/torch/functional.py", line 830, in _unique_consecutive_impl
>     output, inverse_indices, counts = _VF.unique_consecutive(  # type: ignore[attr-defined]

**Comment by julien-c** [comment]
Yes @AsaKal, please comment on https://github.com/pytorch/pytorch/issues/77764

File: src/transformers/pipelines/base.py [hunk 2]
```diff
@@ -760,14 +760,21 @@ def __init__(
         self.feature_extractor = feature_extractor
         self.modelcard = modelcard
         self.framework = framework
-        if is_torch_available() and isinstance(device, torch.device):
-            self.device = device
+        if is_torch_available() and self.framework == "pt":
+            if isinstance(device, torch.device):
+                self.device = device
+            elif type(device) == str:
```
Comment (sgugger) [inline src/transformers/pipelines/base.py:766]:
```suggestion
            elif isinstance(int, str):
```
Will work for both, I believe, and then you can remove the last else (or raise a ValueError there)

File: src/transformers/pipelines/__init__.py [hunk 3]
```diff
@@ -422,6 +422,7 @@ def pipeline(
     revision: Optional[str] = None,
     use_fast: bool = True,
     use_auth_token: Optional[Union[str, bool]] = None,
+    device: Optional[Union[int, str, "torch.device"]] = None,
```
Comment (sgugger) [inline src/transformers/pipelines/__init__.py:425]: Looks like the default should be at `-1` instead of None?

**Comment by julien-c** [comment]
⚠️ You might have to `os.environ["PYTORCH_ENABLE_MPS_FALLBACK"] = "1"` because many operations are still not implemented

also cc @pacman100 for visibility

**Pull request merged**